File: C++/arrays/dynamic_array.cpp
#include <iostream>

using std::cout;
using std::endl;

template <class dtype>
class dynamic_array
{
  public:
    int max = 1;
    int length = 0;
    dtype * data = new dtype [max]; // we begin with just one space

  int push(dtype item) 
  {
    if (length + 1 <= max)
    {
      data[length] = item;
      length++;
    }
    else 
    {
      double_spaces();
      push(item);
    }

    return length;
  }

  int double_spaces()
  {
    dtype dummy_arr[max];
    int i;

    for (i = 0; i < length; i++)
    {
      dummy_arr[i] = data[i];
    }

    delete [] data;
    max = max * 2;
    data = new dtype [max];
    // cout << "doubling spaces...\n";
    for (i = 0; i < length; i++)
    {
      data[i] = dummy_arr[i];
      //cout << data[i] << "\n";
    }

    return max;
  }

  void view_array()
  {
    int i;
    for (i = 0; i < length; i++)
    {
      cout << i << ": " << data[i] << endl;
    }

    return;
  }
};

int main() 
{ 
  dynamic_array<float> arr;

  arr.push(1.1);
  arr.push(2.2);
  arr.push(3.3);
  arr.push(4.4);
  arr.push(5.5);
  
  cout << "length now: " << arr.length << endl;
  arr.view_array();
  cout << "max: " << arr.max;
  
  return 0;
}

File: Python/arrays/dynamic_array.py
class dynamic_array:
  def __init__ (self, dtype, max = 1):
    self.max: int = max
    self.length: int = 0
    self.data: dtype = [None] * max

  def push (self, item):
    if (self.length + 1 <= self.max):
      self.data[self.length] = item
      self.length += 1
    else:
      self.double_spaces()
      self.push(item)

    return self.length

  def double_spaces(self):
    dummy_arr: dtype = [None] * self.max
    i: int = 0

    for i in range(self.length):
      dummy_arr[i] = self.data[i]

    self.max *= 2
    self.data: dtype = [None] * self.max

    for i in range(self.length):
      self.data[i] = dummy_arr[i]

  def view_array(self):
    i: int = 0
    print(f'length: {self.length}')
    for i in range(self.length):
      print(f'{i}: ', self.data[i])
    print(f'max: {self.max}')
    return self.length # O(n)
arr = dynamic_array(int)
arr.push(11)
arr.push(22)
arr.push(33)
arr.push(44)
arr.push(55)
arr.push(66)

arr.view_array()